src/format.py

```diff
@@ -121,42 +128,47 @@ def getSelected(self):
         return self.selected
 
     def getBefore(self):
-        return self.line[0:self.start]
+        return str(self.formattedLine)[0:self.start]
 
     def getAfter(self):
-        return self.line[self.end:]
+        return str(self.formattedLine)[self.end:]
 
     def getMatch(self):
         return self.group
 
     def __str__(self):
-        return self.getBefore() + '||' + self.getMatch(
-        ) + '||' + self.getAfter() + '||' + str(self.num)
-
-    def getStyleForState(self):
-        curses.init_pair(1, curses.COLOR_WHITE, curses.COLOR_RED)
-        curses.init_pair(2, curses.COLOR_WHITE, curses.COLOR_BLUE)
-        curses.init_pair(3, curses.COLOR_WHITE, curses.COLOR_GREEN)
+        return (self.getBefore() + '||' + self.getMatch()
+                + '||' + self.getAfter() + '||' +
+                str(self.num))
 
+    def getDecoratorFormatting(self):
         if self.hovered and self.selected:
-            return curses.color_pair(3)
+            attributes = (curses.COLOR_WHITE, curses.COLOR_RED, 0)
         elif self.hovered:
-            return curses.color_pair(1)
+            attributes = (curses.COLOR_WHITE, curses.COLOR_BLUE, 0)
         elif self.selected:
-            return curses.color_pair(2)
+            attributes = (curses.COLOR_WHITE, curses.COLOR_GREEN, 0)
         else:
-            return curses.A_UNDERLINE
+            attributes = (0,0,FormattedText.UNDERLINE_ATTRIBUTE)
+
+        return FormattedText.getSequenceForAttributes(*attributes)
 
     def getDecorator(self):
         if self.selected:
             return '|===>'
         return ''
 
-    def output(self, stdscr):
+    def output(self, printer):
         decorator = self.getDecorator()
-        before = self.getBefore()
-        after = self.getAfter()
-        middle = ''.join([decorator, self.getMatch()])
+        # inject the decorator and decorated style
+        # into the formatted text and then simply print it out
+        text = copy.deepcopy(self.formattedLine)
```

Comment:
Actually I don't know. I suppose we don't really need to do a copy at all or inject.
Actually, it's much simpler and I will reimplement it this way:

1. Break the line at the beginning and end of the match in __init__, therefore it will be saved by `pickle`, this will happen during `processInput` once only, and actually just save two lines instead of one. Use existing `breakat` and remove the `replace` function.
2. When outputting, output each line segment separately, and also, cache the current `formattedText` representing the decorator. 

It removes the alloc on each output and is also cleaner and removes the need for `replace` functionality. Feels good